omegaconf/listconfig.py

```diff
@@ -101,18 +102,31 @@ def _validate_set(self, key: Any, value: Any) -> None:
                     )
 
         target_type = self._metadata.element_type
+        if target_type in (Any, None):
+            return
+
         value_type = OmegaConf.get_type(value)
-        if is_structured_config(target_type):
-            if (
-                target_type is not None
-                and value_type is not None
-                and not issubclass(value_type, target_type)
-            ):
-                msg = (
-                    f"Invalid type assigned : {type_str(value_type)} is not a "
-                    f"subclass of {type_str(target_type)}. value: {value}"
-                )
-                raise ValidationError(msg)
+
+        def raise_validation_error(
+            value_type: Any, target_type: Any, value: Any
+        ) -> None:
+            msg = (
+                f"Invalid type assigned : {type_str(value_type)} is not a "
+                f"subclass of {type_str(target_type)}. value: {value}"
+            )
+
+            raise ValidationError(msg)
+
+        if (
+            (
+                is_structured_config(value_type)
+                or is_structured_config(target_type)
+                or isinstance(value, ValueNode)
+            )
+            and value_type is not None
+            and not issubclass(value_type, target_type)
+        ):
+            raise_validation_error(value_type, target_type, value)
```

Comment:
Can you explain what new cases this is dealing with?
it's significantly more complicated than before and looking at it I suspect it's actually too broad.

Reply:
Well if we have a structured config the way to check is with `issubclass` so that's one case. 

The case with ValueNodes is because the same reason with other PRs where we receive a node instead of a primitive value and, this is where we can check.

Reply:
Trying to parse the logic here:
```
        if (
            (
                is_structured_config(value_type)
                or is_structured_config(target_type)
                or isinstance(value, ValueNode)
            )
```

```
If:
  input is structured config (e.g an instance of or a dataclass)
  or element type is a structured config
  or value we are assigning a value node
```
This is already lumping multiple unrealted things together.

```
            and value_type is not None
            and not issubclass(value_type, target_type)
```

and the value type is not a subclass of the element type.
=> fail.


To me a better way to express this sounds like:
```
if value is a structured config:
  if element type is a structured config and value type is not a subclass
    => fail
```

I don't know what the input being a value node has to do with it, shouldn't the check be performed on the unboxed input?

Reply:
> Trying to parse the logic here:
> 
> ```
>         if (
>             (
>                 is_structured_config(value_type)
>                 or is_structured_config(target_type)
>                 or isinstance(value, ValueNode)
>             )
> ```
> 
> ```
> If:
>   input is structured config (e.g an instance of or a dataclass)
>   or element type is a structured config
>   or value we are assigning a value node
> ```
> 
> This is already lumping multiple unrealted things together.
> 
> ```
>             and value_type is not None
>             and not issubclass(value_type, target_type)
> ```
> 
> and the value type is not a subclass of the element type.
> => fail.
> 
> To me a better way to express this sounds like:
> 
> ```
> if value is a structured config:
>   if element type is a structured config and value type is not a subclass
>     => fail
> ```

I agree.
> 
> I don't know what the input being a value node has to do with it, shouldn't the check be performed on the unboxed input?

ValueNodes are a bit tricky because of implicit conversion. I think this process could be much more simplified with :
```
if value is not a subscripted container type and not subclass..
```

But since value nodes implicitly convert values we cannot. Therefore, we arrive to a point where in the following case 
```python
@attr.s(auto_attribs=True)
class ComplexDict:
    dict: Dict[str, Dict[str, List[int]]] = {"foo": {"var": [1, 2]}}

obj = {
          "dict": {"foo2": {"var2": ["invalid", 4]}},
      }
c = OmegaConf.structured(ComplexDict)
OmegaConf.merge(c, obj) # no error
```
This doesn't raise an error because "invalid" is an StringNode which cannot be checked its value without defining an explicit check in `_validate_set` and, if we check with the unboxed value(str literal) it can cause regression problems since, for example, the int value `4` can be StringNode too(`issubclass(int, str)` will fail but this is ok in omegaconf).

I don't know if I explained this correctly, I hope this one clarifies my concerns.